state-chain/pallets/cf-account-roles/src/lib.rs

```diff
@@ -152,6 +186,84 @@ pub mod pallet {
 			Self::deposit_event(Event::VanityNameSet { account_id, name });
 			Ok(())
 		}
+
+		/// Derives a sub-account by the given origin account id and a sub-account index.
+		/// Stores the account id against the sub-account index and origin account id.
+		///
+		/// The sub account is unmutual associated with the origin/parent account and the creation
+		/// fails if the sub-account already exists for the given index. It's possible to derive 256
+		/// sub-accounts per parent account. The parent account can execute calls on behalf of the
+		/// sub-account (see as_sub_account).
+		///
+		/// Events:
+		/// - `SubAccountCreated` if the sub-account is created.
+		///
+		/// Errors:
+		/// - `SubAccountAlreadyExists` if the sub-account already exists for the given index.
+		/// - `SubAccountIdDerivationFailed` if the sub-account id derivation fails.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::derive_sub_account())]
+		pub fn derive_sub_account(
+			origin: OriginFor<T>,
+			sub_account_index: SubAccountIndex,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			ensure!(
+				!SubAccounts::<T>::contains_key(&account_id, sub_account_index),
+				Error::<T>::SubAccountAlreadyExists
+			);
+			let sub_account_id: T::AccountId = Decode::decode(&mut TrailingZeroInput::new(
+				(*b"chainflip/subaccount", account_id.clone(), sub_account_index)
+					.blake2_256()
+					.as_ref(),
+			))
+			.map_err(|_| Error::<T>::SubAccountIdDerivationFailed)?;
+			SubAccounts::<T>::insert(&account_id, sub_account_index, &sub_account_id);
```

Comment:
A subaccount needs to be a 'real' account, and should be created in the same way that accounts are created in the funding pallet (maybe we expose an internal 'fund_account' API. Then we also don't need the `SubAccounts` storage (we just derive the account and check if it exists, like any other account). It still might be useful to store *some* mapping to allow us to aggregate eg. balances from the subaccount to the parent in the rpcs. 

Using the current implementation you can't, for example, use rebalance to add funds to the account because the account doesn't exist. 

I think this implies that the subaccount needs to be created with some funds from the parent (basically like the `funded` extrinsic in the funding pallet).